File: mobile/lib/services/worker_endpoints.dart
// ═══════════════════════════════════════════════════════════════
//  worker_endpoints.dart — سجل نقاط نهاية Worker مع تبديل تلقائي
//  ✅ (2026-09-09) طلب المستخدم: «دعم عدّة عناوين Worker مع تبديل
//  تلقائي (workers.dev + نطاق مخصّص) — يضع المستخدم نطاقه فيعمل
//  التطبيق عبره تلقائياً».
// ═══════════════════════════════════════════════════════════════
//
//  السياق الهندسي: شبكات اليمن تحجب *.workers.dev (فلترة SNI/IP/DNS).
//  لا يوجد إصلاح برمجي بحت يتجاوز حجب SNI بثبات — الحل الصحيح نطاق
//  مخصّص مربوط بنفس الـ Worker (SNI مسموح). هذا السجل يدير المرشحين:
//
//    1. النطاق المخصّص (يضبطه المستخدم من إعدادات المزامنة) — أولوية
//       دائمة ما دام موجوداً: وجوده بحد ذاته إشارة أن المدمج محجوب.
//    2. النطاق المدمج workers.dev (--dart-define CLOUDFLARE_WORKER_URL).
//
//  ويثبّت آخر نقطة نجحت (sticky) محفوظة في SharedPreferences — كل
//  URL لاحق يُبنى مباشرة على الفائز دون دفع كلفة تجريب كل طلب. الفشل
//  يُنزّل الفائز مؤقتاً ويمرّر الطلب للمرشح التالي (يُنفَّذ الفعلي
//  داخل ResilientHttpClient عبر endpointPlanner — نجاح/فشل يرجع هنا
//  عبر reportSuccess/reportFailure).

import 'dart:async';

import 'package:flutter/foundation.dart';
import 'package:shared_preferences/shared_preferences.dart';

import '../utils/debug_log.dart';
import '../utils/env.dart';

class WorkerEndpoints {
  WorkerEndpoints._();

  /// مفتاح النطاق المخصّص في SharedPreferences.
  static const String customUrlKey = 'cf_custom_worker_url';

  /// مفتاح آخر نقطة نهاية نجحت (sticky) — يُستعاد عند الإطلاق.
  static const String activeUrlKey = 'cf_worker_active_url';

  static String? _customUrl;
  static String? _activeOverride;
  static bool _loaded = false;
  static SharedPreferences? _prefsRef;

  /// النقطة المدمجة من بيئة البناء (workers.dev).
  static String get builtin => Env.cloudflareWorkerUrl;

  /// هل حُمّل السجل من التفضيلات؟ (يستدعى من main مبكراً).
  static bool get isLoaded => _loaded;

  /// النطاق المخصّص الحالي (null = غير مضبوط).
  static String? get custom => _customUrl;
  static bool get hasCustom => _customUrl != null;

  /// تطبيع إدخال المستخدم: «mydomain.com» أو «https://mydomain.com/»
  /// أو مع منفذ — يعيد URL نظيفاً https://host[:port] بلا مسار/استعلام.
  /// يعيد null عند إدخال فارغ، ويرمي [FormatException] عند إدخال فاسد.
  /// عامة عمداً — تستخدمها شاشة الإعدادات للتحقق قبل الفحص.
  static String? normalizeCustomUrl(String? raw) {
    final trimmed = raw?.trim() ?? '';
    if (trimmed.isEmpty) return null;
    var text = trimmed;
    if (!text.contains('://')) text = 'https://$text';
    final Uri uri;
    try {
      uri = Uri.parse(text);
    } on FormatException {
      throw FormatException('رابط غير صالح: $trimmed');
    }
    final host = uri.host.toLowerCase();
    if (host.isEmpty || host.contains(' ') || !host.contains('.')) {
      throw FormatException('اسم نطاق غير صالح: $trimmed');
    }
    if (uri.scheme != 'https') {
      throw FormatException('يجب أن يكون الرابط https: $trimmed');
    }
    if (uri.userInfo.isNotEmpty ||
        (uri.path.isNotEmpty && uri.path != '/') ||
        uri.hasQuery) {
      throw FormatException('أدخل النطاق الجذر فقط بدون مسار: $trimmed');
    }
    return uri.port == 443 || uri.port == 0
        ? 'https://$host'
        : 'https://$host:${uri.port}';
  }

  /// تحميل الحالة المحفوظة — يستدعى مرة واحدة مبكراً في main().
  /// fail-open: أي فشل يعني الاستمرار بالنطاق المدمج.
  static Future<void> load({SharedPreferences? prefs}) async {
    try {
      final sp = prefs ?? await SharedPreferences.getInstance();
      _prefsRef = sp;
      _customUrl = _sanitize(sp.getString(customUrlKey));
      _activeOverride = _sanitize(sp.getString(activeUrlKey));
      // نطاق مخصّص موجود = إشارة أن المدمج غير موثوق (سبب وضعه أصلاً).
      // ابدأ الجلسة عليه؛ الفشل يصحّح التثبيت تلقائياً عبر التدوير،
      // والنجاح على المدمج يعيده sticky خلال الجلسة.
      if (_customUrl != null && _activeOverride == builtin) {
        _activeOverride = _customUrl;
      }
    } catch (e) {
      dwarn(() => 'WorkerEndpoints.load failed (fail-open to builtin): $e');
      _activeOverride = null;
    } finally {
      _loaded = true;
    }
  }

  /// العنوان الفعّال الذي تُبنى عليه كل روابط الـ Worker.
  /// قبل التحميل يعيد المدمج (سلوك مطابق للسابق — لا كسر للاختبارات).
  static String get active {
    if (!_loaded) return builtin;
    final sticky = _sanitize(_activeOverride);
    if (sticky != null && _isRegisteredBase(sticky)) return sticky;
    return _customUrl ?? builtin;
  }

  /// هل [uri] يشير إلى إحدى نقاط الـ worker المسجلة؟
  static bool isWorkerEndpoint(Uri uri) {
    final key = _hostKey(uri);
    for (final base in _orderedBases) {
      final u = Uri.tryParse(base);
      if (u != null && _hostKey(u) == key) return true;
    }
    return false;
  }

  /// مرشحو نقاط النهاية لطلب نحو [requestUrl]:
  /// - طلب خارج عائلة نقاط الـ worker (api.cloudflare.com مثلًا) →
  ///   قائمة وحيدة = العنوان نفسه (لا تدوير — عزل كامل).
  /// - طلب worker → قاعدة العنوان الحالي أولاً ثم بقية المرشحين
  ///   بترتيب [custom ثم builtin] بعد إزالة التكرار.
  static List<Uri> candidatesFor(Uri requestUrl) {
    if (!isWorkerEndpoint(requestUrl)) {
      return <Uri>[requestUrl];
    }
    final result = <Uri>[];
    void add(String base) {
      final u = Uri.tryParse(base);
      if (u == null) return;
      final key = _hostKey(u);
      if (result.any((r) => _hostKey(r) == key)) return;
      result.add(u);
    }

    // العنوان الذي بُني عليه الطلب أولاً (محاولة صفر كلفة إعادة كتابة).
    add(requestUrl.toString());
    if (_customUrl != null) add(_customUrl!);
    add(builtin);
    return result;
  }

  /// نجاح طلب على [base] — يثبّته للطلبات اللاحقة (ويعاد استعادته
  /// عند الإطلاق القادم). الحفظ fire-and-forget — لا يعطّل مسار الطلب.
  static void reportSuccess(Uri base) {
    final normalized = _sanitize(_toBaseUrl(base));
    if (normalized == null) return;
    if (!_isRegisteredBase(normalized)) return;
    if (_sanitize(_activeOverride) == normalized) return;
    _activeOverride = normalized;
    dlog(() => '🔗 WorkerEndpoints: sticky endpoint → $normalized');
    final sp = _prefsRef;
    if (sp != null) {
      unawaited(
        sp.setString(activeUrlKey, normalized).catchError((Object _) => false),
      );
    }
  }

  /// فشل طلب على [base] — يُنزّله من الصدارة مؤقتاً (in-memory فقط:
  /// التحميل عند الإطلاق يفرض المخصّص أولاً عند وجوده، والنجاح وحده
  /// ما يُثبَّت — لن يعلق المستخدم على نقطة ميتة عبر جلسة كاملة).
  static void reportFailure(Uri base) {
    final normalized = _sanitize(_toBaseUrl(base));
    if (normalized == null) return;
    if (_sanitize(_activeOverride) != normalized) return;
    // انتقل للمرشح التالي: فشل المخصّص → المدمج، وفشل المدمج → المخصّص.
    _activeOverride = _customUrl != null && normalized == builtin
        ? _customUrl
        : (normalized == _customUrl ? builtin : null);
    if (_sanitize(_activeOverride) == normalized) _activeOverride = null;
    dwarn(() => '⚠️ WorkerEndpoints: demoted $normalized (failure)');
  }

  /// ضبط/مسح النطاق المخصّص من الإعدادات.
  /// يعيد الرابط المُطبَّع المحفوظ (null عند المسح).
  /// يرمي [FormatException] على إدخال فاسد (لا يلمس الحالة المحفوظة).
  static Future<String?> setCustomUrl(
    String? raw, {
    SharedPreferences? prefs,
  }) async {
    final normalized = normalizeCustomUrl(raw);
    final sp = prefs ?? _prefsRef ?? await SharedPreferences.getInstance();
    _prefsRef = sp;
    if (normalized == null) {
      await sp.setString(customUrlKey, '');
    } else {
      await sp.setString(customUrlKey, normalized);
    }
    _customUrl = normalized;
    // المخصّص الجديد يتقدم فوراً (المستخدم وضعه لسبب) — والمسح يعيد
    // للمدمج. يُحفظ sticky الجديد ليُستعاد عند الإطلاق القادم.
    _activeOverride = normalized ?? builtin;
    await sp.setString(activeUrlKey, _activeOverride!);
    return normalized;
  }

  // ─── internals ───────────────────────────────────────────────

  /// المرشحون المسجلون بترتيب الأولوية: المخصّص ثم المدمج.
  static List<String> get _orderedBases => <String>[
    if (_customUrl != null) _customUrl!,
    builtin,
  ];

  static bool _isRegisteredBase(String base) {
    for (final b in _orderedBases) {
      final u = Uri.tryParse(b);
      final bu = Uri.tryParse(base);
      if (u != null && bu != null && _hostKey(u) == _hostKey(bu)) return true;
    }
    return false;
  }

  static String? _sanitize(String? raw) {
    if (raw == null) return null;
    final trimmed = raw.trim();
    if (trimmed.isEmpty) return null;
    final uri = Uri.tryParse(trimmed);
    if (uri == null || uri.host.isEmpty) return null;
    return trimmed;
  }

  /// قاعدة نقية scheme://host[:port] — تقص أي مسار/استعلام قد يحمله
  /// الـUri القادم من تقرير نجاح/فشل (المرشح الأول هو الطلب كاملاً).
  /// بدون هذا كان sticky قد يحمل مسار طلب عابر فيفسد بناء الروابط.
  static String _toBaseUrl(Uri u) {
    final port = u.port;
    final hasExplicitPort =
        port != 0 &&
        !((u.scheme == 'https' && port == 443) ||
            (u.scheme == 'http' && port == 80));
    return hasExplicitPort
        ? '${u.scheme}://${u.host.toLowerCase()}:$port'
        : '${u.scheme}://${u.host.toLowerCase()}';
  }

  /// هوية مضيف مستقرة: host:port مع افتراض المنفذ من المخطط.
  static String _hostKey(Uri u) {
    final port = u.port == 0 ? (u.scheme == 'https' ? 443 : 80) : u.port;
    return '${u.host.toLowerCase()}:$port';
  }

  /// Test-only: تفريغ الحالة الساكنة.
  @visibleForTesting
  static void resetForTests() {
    _customUrl = null;
    _activeOverride = null;
    _loaded = false;
    _prefsRef = null;
  }
}

File: mobile/android/app/src/main/kotlin/com/marina/marina/data/remote/WorkerEndpoints.kt
package com.marina.marina.data.remote

import android.content.Context
import dagger.hilt.android.qualifiers.ApplicationContext
import java.net.URI
import javax.inject.Inject
import javax.inject.Singleton

/**
 * ✅ (2026-09-24) سجل نقاط نهاية الـ Worker مع تبديل تلقائي — نقل مباشر من
 * Flutter (worker_endpoints.dart، فرع feat/cloudflare-sync-execution).
 *
 * السياق الهندسي (نفس تعليق Flutter): شبكات اليمن تحجب *.workers.dev
 * (فلترة SNI/IP/DNS). لا إصلاح برمجي بحت يتجاوز حجب SNI بثبات — الحل
 * الصحيح نطاق مخصّص مربوط بنفس الـ Worker (SNI مسموح). هذا السجل يدير
 * المرشحين:
 *
 *    1. النطاق المخصّص (يضبطه المستخدم من شاشة الدخول) — أولوية دائمة
 *       ما دام موجوداً: وجوده بحد ذاته إشارة أن المدمج محجوب.
 *    2. النطاق المدمج workers.dev (CloudflareConfig.BUILTIN_WORKER_URL).
 *
 * ويثبّت آخر نقطة نجحت (sticky) محفوظة في SharedPreferences — الفشل يُنزّل
 * الفائز مؤقتاً ويمرّر الطلب للمرشح التالي (ينفَّذ فعلياً داخل
 * [WorkerFailoverInterceptor] — نجاح/فشل يرجع هنا عبر reportSuccess/reportFailure).
 */
@Singleton
class WorkerEndpoints @Inject constructor(
    @ApplicationContext private val context: Context
) {
    companion object {
        /** مفتاح النطاق المخصّص في SharedPreferences (نفس مفتاح Flutter). */
        const val CUSTOM_URL_KEY = "cf_custom_worker_url"

        /** مفتاح آخر نقطة نهاية نجحت (sticky) — يُستعاد عند الإطلاق. */
        const val ACTIVE_URL_KEY = "cf_worker_active_url"

        @Volatile private var initialized = false
        @Volatile private var customUrl: String? = null
        @Volatile private var activeOverride: String? = null
        private val lock = Any()

        /** Test-only: تفريغ الحالة الساكنة (توازي resetForTests في Flutter). */
        fun resetForTests() {
            synchronized(lock) {
                initialized = false
                customUrl = null
                activeOverride = null
            }
        }
    }

    private val prefs by lazy {
        context.getSharedPreferences("marina_cloudflare_prefs", Context.MODE_PRIVATE)
    }

    /** النقطة المدمجة من بيئة البناء (workers.dev). */
    val builtin: String get() = CloudflareConfig.BUILTIN_WORKER_URL

    /** تحميل الحالة المحفوظة — يستدعى مرة واحدة مبكراً (fail-open للمدمج). */
    fun loadIfNeeded() {
        synchronized(lock) {
            if (initialized) return
            val custom = sanitize(prefs.getString(CUSTOM_URL_KEY, null))
            val sticky = sanitize(prefs.getString(ACTIVE_URL_KEY, null))
            customUrl = custom
            // نطاق مخصّص موجود = إشارة أن المدمج غير موثوق (سبب وضعه أصلاً).
            // ابدأ الجلسة عليه؛ الفشل يصحّح التثبيت تلقائياً عبر التدوير.
            activeOverride = if (custom != null && sticky == builtin) custom else sticky
            initialized = true
        }
    }

    /** العنوان الفعّال الذي تُبنى عليه كل روابط الـ Worker. */
    val active: String
        get() {
            loadIfNeeded()
            val sticky = sanitize(activeOverride)
            if (sticky != null && isRegisteredBase(sticky)) return sticky
            return customUrl ?: builtin
        }

    /** النطاق المخصّص الحالي (null = غير مضبوط). */
    val custom: String?
        get() {
            loadIfNeeded()
            return customUrl
        }

    val hasCustom: Boolean get() = custom != null

    /**
     * تطبيع إدخال المستخدم: «mydomain.com» أو «https://mydomain.com/» →
     * URL نظيف https://host[:port] بلا مسار/استعلام. يعيد null عند إدخال
     * فارغ، ويرمي [IllegalArgumentException] عند إدخال فاسد (نفس عقود
     * normalizeCustomUrl في Flutter: FormatException).
     */
    fun normalizeCustomUrl(raw: String?): String? {
        val trimmed = raw?.trim().orEmpty()
        if (trimmed.isEmpty()) return null
        var text = trimmed
        if (!text.contains("://")) text = "https://$text"
        val uri = try {
            URI(text)
        } catch (e: Exception) {
            throw IllegalArgumentException("رابط غير صالح: $trimmed")
        }
        val host = (uri.host ?: "").lowercase()
        if (host.isEmpty() || host.contains(' ') || !host.contains('.')) {
            throw IllegalArgumentException("اسم نطاق غير صالح: $trimmed")
        }
        if (uri.scheme != "https") {
            throw IllegalArgumentException("يجب أن يكون الرابط https: $trimmed")
        }
        val path = uri.path ?: ""
        if (uri.userInfo != null || (path.isNotEmpty() && path != "/") || uri.rawQuery != null) {
            throw IllegalArgumentException("أدخل النطاق الجذر فقط بدون مسار: $trimmed")
        }
        return if (uri.port == 443 || uri.port == -1) "https://$host" else "https://$host:${uri.port}"
    }

    /**
     * ضبط/مسح النطاق المخصّص من شاشة الدخول. يعيد الرابط المُطبَّع المحفوظ
     * (null عند المسح). يرمي [IllegalArgumentException] على إدخال فاسد
     * (لا يلمس الحالة المحفوظة).
     */
    fun setCustomUrl(raw: String?): String? {
        loadIfNeeded()
        val normalized = normalizeCustomUrl(raw)
        synchronized(lock) {
            if (normalized == null) {
                prefs.edit().putString(CUSTOM_URL_KEY, "").apply()
            } else {
                prefs.edit().putString(CUSTOM_URL_KEY, normalized).apply()
            }
            customUrl = normalized
            // المخصّص الجديد يتقدم فوراً (المستخدم وضعه لسبب) — والمسح يعيد
            // للمدمج. يُحفظ sticky الجديد ليُستعاد عند الإطلاق القادم.
            activeOverride = normalized ?: builtin
            prefs.edit().putString(ACTIVE_URL_KEY, activeOverride!!).apply()
        }
        return normalized
    }

    /** مرشحو نقاط النهاية لطلب نحو [requestUrl]: الطلب نفسه أولاً ثم بقية المرشحين. */
    fun candidatesFor(requestUrl: URI): List<URI> {
        loadIfNeeded()
        if (!isWorkerEndpoint(requestUrl)) return listOf(requestUrl)
        val result = mutableListOf<URI>()
        fun add(base: String) {
            val u = runCatching { URI(base) }.getOrNull() ?: return
            if (result.none { hostKey(it) == hostKey(u) }) result.add(u)
        }
        add(toBaseUrl(requestUrl).toString())
        customUrl?.let { add(it) }
        add(builtin)
        return result
    }

    /** نجاح طلب على [base] — يثبّته للطلبات اللاحقة (sticky). */
    fun reportSuccess(base: URI) {
        loadIfNeeded()
        val normalized = sanitize(toBaseUrl(base).toString()) ?: return
        if (!isRegisteredBase(normalized)) return
        synchronized(lock) {
            if (sanitize(activeOverride) == normalized) return
            activeOverride = normalized
            prefs.edit().putString(ACTIVE_URL_KEY, normalized).apply()
        }
    }

    /** فشل طلب على [base] — يُنزّله من الصدارة مؤقتاً (in-memory فقط). */
    fun reportFailure(base: URI) {
        loadIfNeeded()
        val normalized = sanitize(toBaseUrl(base).toString()) ?: return
        synchronized(lock) {
            if (sanitize(activeOverride) != normalized) return
            // انتقل للمرشح التالي: فشل المخصّص → المدمج، وفشل المدمج → المخصّص.
            activeOverride = if (customUrl != null && normalized == builtin) {
                customUrl
            } else if (normalized == customUrl) {
                builtin
            } else {
                null
            }
            if (sanitize(activeOverride) == normalized) activeOverride = null
        }
    }

    /** هل [uri] يشير إلى إحدى نقاط الـ worker المسجلة؟ */
    fun isWorkerEndpoint(uri: URI): Boolean {
        loadIfNeeded()
        val key = hostKey(uri)
        return orderedBases().any { hostKey(URI(it)) == key }
    }

    // ─── internals ───────────────────────────────────────────────

    private fun orderedBases(): List<String> = listOfNotNull(customUrl, builtin)

    private fun isRegisteredBase(base: String): Boolean {
        val bu = runCatching { URI(base) }.getOrNull() ?: return false
        return orderedBases().any { hostKey(URI(it)) == hostKey(bu) }
    }

    private fun sanitize(raw: String?): String? {
        if (raw == null) return null
        val trimmed = raw.trim()
        if (trimmed.isEmpty()) return null
        val uri = runCatching { URI(trimmed) }.getOrNull() ?: return null
        if (uri.host.isNullOrEmpty()) return null
        return trimmed
    }

    /** قاعدة نقية scheme://host[:port] — تقص أي مسار/استعلام عابر. */
    private fun toBaseUrl(u: URI): URI {
        val port = if (u.port == -1) -1 else u.port
        return URI(u.scheme, u.userInfo, u.host.lowercase(), port, null, null, null)
    }

    /** هوية مضيف مستقرة: host:port مع افتراض المنفذ من المخطط. */
    private fun hostKey(u: URI): String {
        val port = if (u.port == -1) if (u.scheme == "https") 443 else 80 else u.port
        return "${u.host.lowercase()}:$port"
    }
}
